### PR TITLE
Weapon Analysis Printing

### PR DESCRIPTION
**Pull request by Geevies**
* Weapon Analyzers can now print reports about their inserted items via a button at the bottom of the UI.

This also fixes a runtime with an asset of an item that doesn't exist being sent to the UI.

### CODE REVIEW

**Comment by Geevies** [comment]
!review

File: code/modules/research/weaponsanalyzer.dm [hunk 1]
```diff
@@ -252,9 +255,123 @@
...
 
-	var/icon/Icon_used
 	if(item)
-		Icon_used = new /icon(item.icon, item.icon_state)
-	ui.add_asset("icon", Icon_used) 
-	ui.send_asset("icon")
-	ui.open()
\ No newline at end of file
+		var/icon/Icon_used
+		if(item)
```
Comment (Amunak) [inline code/modules/research/weaponsanalyzer.dm]: The item check seems duplicate.

**Review by Amunak — APPROVED** [review]
I really like this, though I think it would be better to create a proc on item like  `/obj/item/proc/get_print_analysis(var/obj/machinery/printer)` that would gather this kind of information so that it's (1) potentially useful elsewhere and (2) overridable by sub-types. I would also let that proc decide if it want to get analyzed and what / how much information it gives based on the printer argument.

Other than that I really like it.

**Comment by Geevies** [comment]
> 
> 
> I really like this, though I think it would be better to create a proc on item like `/obj/item/proc/get_print_analysis(var/obj/machinery/printer)` that would gather this kind of information so that it's (1) potentially useful elsewhere and (2) overridable by sub-types. I would also let that proc decide if it want to get analyzed and what / how much information it gives based on the printer argument.

I think that's doable, but I doooon't think that's really necessary for this. If someone wants to do that for another machine, creating that should be fairly easy.

File: code/modules/research/weaponsanalyzer.dm [hunk 2]
```diff
@@ -252,9 +255,121 @@
...
+		dat += "Damage: [item.force]<br>"
+		dat += "Damage Type: [item.damtype]<br>"
+		dat += "Sharp: [item.sharp ? "yes" : "no"]<br>"
+		dat += "Dismemberment: [item.edge ? "likely to dismember" : "unlikely to dismember"]<br>"
+		dat += "Penetration: [item.armor_penetration]<br>"
+		dat += "Throw Force: [item.throwforce]<br>"
+		if(istype(item, /obj/item/melee/energy))
+			var/obj/item/melee/energy/E_item = item
+			dat += "Active Damage: [item.force]<br>"
+			dat += "Active Throw Force: [item.force]<br>"
```
Comment (NonQueueingMatt) [inline code/modules/research/weaponsanalyzer.dm:369]: This looks wrong. Throw force is its own var and I'm sure that active damage is not base force.

**Review by NonQueueingMatt — CHANGES_REQUESTED** [review]
I have to agree with amunak that these kinds of things should start getting offloaded to the object itself rather than a nebulous proc on the machine.

**Review by JohnWildkins — APPROVED** [review]

File: code/modules/projectiles/guns/energy.dm [hunk 3]
```diff
@@ -151,3 +151,35 @@
...
+		. += "Shrapnel Type: [shrapnel.name]<br>"
+	. += "Armor Penetration: [initial(P.armor_penetration)]%<br>"
+
+	if(secondary_projectile_type)
+		var/obj/item/projectile/P_second = secondary_projectile_type
+		. += "<br><b>Secondary Projectile</b><br>"
+		. += "Damage: [initial(P_second.damage)]<br>"
+		. += "Damage Type: [initial(P_second.damage_type)]<br>"
+		. += "Blocked by Armor Type: [initial(P_second.check_armor)]<br>"
+		. += "Stuns: [initial(P_second.stun) ? "true" : "false"]<br>"
```
Comment (NonQueueingMatt) [inline code/modules/projectiles/guns/energy.dm:178]: Handle this on the projectile.

File: code/modules/projectiles/guns/projectile.dm [hunk 4]
```diff
@@ -268,3 +268,19 @@
...
+
+/obj/item/gun/projectile/get_print_info()
+	. = ""
+	var/obj/item/ammo_casing/casing = new ammo_type
+	var/obj/item/projectile/P = new casing.projectile_type
+	. += "Max Shots: [max_shells]<br>"
+	. += "Damage: [initial(P.damage)]<br>"
+	. += "Damage Type: [initial(P.damage_type)]<br>"
+	. += "Blocked by Armor Type: [initial(P.check_armor)]<br>"
+	. += "Stuns: [initial(P.stun) ? "true" : "false"]<br>"
```
Comment (NonQueueingMatt) [inline code/modules/projectiles/guns/projectile.dm:280]: Ditto.

**Review by NonQueueingMatt — CHANGES_REQUESTED** [review]

**Review by NonQueueingMatt — APPROVED** [review]

**Pull request merged**